### ISSUE TITLE
Make alias required for v-repeat 

### ISSUE THREAD

**Issue by yyx990803** [yyx990803/vue#1168]
### Problem

Currently there are two ways to write a `v-repeat`: with or without an alias. The without-alias syntax has been supported for a long time, but it has a few issues:

1. It makes it unclear whether a property belongs in the current repeat scope or in the parent scope.
2. We have to use the magical `$value` to refer to primitive values.
3. Allowing two different syntaxes complicates internal implementation of `v-repeat`.
4. When a component is used without alias, the repeated object is passed into the component as `$data`; if the component also has a default `data` function, the fields returned in it will be merged into the passed in object, essentially mutating the raw data in the Array.

### Proposal

Make alias required.

**Comment by Boydbueno** [comment]
Agreed

**Comment by holic** [comment]
:+1:

**Comment by cheapsteak** [comment]
How would this affect iterating through an object?

**Comment by yyx990803** [comment]
@cheapsteak yep. The only different is `$key` instead of `$index`

**Comment by cheapsteak** [comment]
Should there be a way to alias $key as well for nested repeats?

```js
var things = {
  thing1: {
    subthing1: {},
    subthing2: {},
    subthing3: {},
  },
  thing2: { /*..*/ },
  thing3: { /*..*/ }
}
```

```vue
<template v-repeat="thing in things">
  <template v-repeat="subthing in thing">
    {{$key}}
  </template>
</template>
```

**Comment by yyx990803** [comment]
Yeah, currently you'd have to do `$parent.$key`. I really want to avoid complicating the `v-repeat` syntax into something of its own though.

**Comment by simplesmiler** [comment]
@yyx990803 maybe v-repeat should not have supported objects in the first place, because there is no single way to order object keys.

**Comment by mark-hahn** [comment]
> there is no single way to order object keys.

Does a list on a web page always have to be ordered?  What if you are just
showing a series of images that have no order?

**Comment by yyx990803** [comment]
@simplesmiler `v-repeat="val in object | orderBy '$key'"`?

**Comment by simplesmiler** [comment]
@mark-hahn DOM elements are certainly ordered.

@yyx990803 `v-repeat="val in (object | orderBy '$key')"` seems to me like repeat over an array, which was the point of my message (parenthesis are imaginary, not as part of syntax).

**Comment by mark-hahn** [comment]
> DOM elements are certainly ordered.

But the contents of the DOM don't have to be.  I gave the example where I
don't care about the order of images shown.

**Comment by simplesmiler** [comment]
I guess I should have opened a new issue for this, but oh well.

@mark-hahn data can be unordered for sure, and the way the user perceives data representation may be unordered, but the representation we as developers are working with is always ordered. And I think we should not abstract too far away from this fact.

There are different scenarios for 'things without order':
- Sometimes we do not care about particular order of items, because the visual representation is independent of the order (absolutely positioned points on a scatterplot).
- Sometimes we do not care about particular order of items, but want the ordering to be 'stable', so that after refresh or insertion of a new item other items don't jump around too much (participants of the discussion).
- Sometimes we want items to be ordered quasi-randomly, so that every item is given equal exposure.
- Sometimes items do not have inherent order, but may be ordered given criteria (rows in a table with sortable columns).

I understand, that all of the above may be implemented using current syntax. But I think that repeating over object keys **with** `orderBy` is no different to repeating over array. And repeating over object keys **without** `orderBy`:

1. Gives false sense of being stably or quasi-randomly ordered when it's not.
2. May be improved perfomance-wise for truly order-independent data.

@yyx990803, am I wrong in the assumption, that for truly order-independent data there are less modifications to be made to the DOM when data changes?

**Comment by JosephSilber** [comment]
@simplesmiler in ES6, object key order has been defined.

**Comment by yyx990803** [comment]
@simplesmiler the implied behavior of repeating over object keys without orderBy is that the order will be the same as the property enumeration order implemented by the browser's JavaScript engine (same as the order of `for (key in obj)`). In my opinion this is a reasonable assumption and separating object iteration into a different syntax / implementation only complicates the issue.

**Comment by simplesmiler** [comment]
@JosephSilber thanks, was not aware of that. Do you think transpilation or aggressive minification or (like babel and closure compiler) can break this order, so that original and minified scripts behave differently?

@yyx990803 this is fair, doing as javascript does. But still, do you think it's possible to write a special directive for unordered data, that would outperform general case `v-repeat`?

**Comment by JosephSilber** [comment]
No browser has implemented this yet, so there's no way to know.

**Comment by yyx990803** [comment]
Closed in favor of #1200.
